category/execution/ethereum/precompiles_impl.cpp

```diff
@@ -182,11 +184,152 @@ uint64_t blake2bf_gas_cost(byte_string_view const input)
 
 EXPLICIT_EVM_TRAITS(blake2bf_gas_cost);
 
+static void
+right_pad(std::basic_string<uint8_t> &str, size_t const min_size) noexcept
+{
+    if (str.length() < min_size) {
+        str.resize(min_size, '\0');
+    }
+}
+
+static runtime::uint256_t
+mult_complexity_eip198(runtime::uint256_t const &x) noexcept
+{
+    runtime::uint256_t const x_squared{x * x};
+    if (x <= 64) {
+        return x_squared;
+    }
+    else if (x <= 1024) {
+        return (x_squared >> 2) + 96 * x - 3072;
+    }
+    else {
+        return (x_squared >> 4) + 480 * x - 199680;
+    }
+}
+
+static runtime::uint256_t
+mult_complexity_eip2565(runtime::uint256_t const &max_length) noexcept
+{
+    runtime::uint256_t const words{(max_length + 7) >> 3}; // ceil(max_length/8)
+    return words * words;
+}
+
+static runtime::uint256_t
+mult_complexity_eip7883(runtime::uint256_t const &max_length) noexcept
+{
+    runtime::uint256_t const words{(max_length + 7) >> 3}; // ceil(max_length/8)
+    if (max_length > 32) {
+        return 2 * words * words;
+    }
+    else {
+        return 16;
+    }
+}
+
+template <Traits traits>
+runtime::uint256_t
+expmod_iteration_count(runtime::uint256_t exp_len256, size_t bit_len) noexcept
+{
+    runtime::uint256_t adjusted_exponent_len{0};
+    if (exp_len256 > 32) {
+        constexpr runtime::uint256_t exp_mult{
+            traits::evm_rev() < EVMC_OSAKA ? 8u : 16u};
+        adjusted_exponent_len = exp_mult * (exp_len256 - 32);
+    }
+    if (bit_len > 1) {
+        adjusted_exponent_len = adjusted_exponent_len +
+                                static_cast<runtime::uint256_t>(bit_len - 1);
+    }
+
+    return std::max(adjusted_exponent_len, runtime::uint256_t{1});
+}
+
 template <Traits traits>
-uint64_t expmod_gas_cost(byte_string_view const input)
+constexpr uint64_t expmod_min_gas()
 {
-    return silkpre_expmod_gas(
-        input.data(), input.size(), static_cast<int>(traits::evm_rev()));
+    if (traits::evm_rev() < EVMC_BERLIN) {
+        return 0; // Prior to Berlin, minimum gas is 0.
+    }
+    else if (traits::evm_rev() < EVMC_OSAKA) {
+        return 200; // From Berlin to pre-Osaka (Prague), minimum gas is 200.
+    }
+    else {
+        return 500; // From Osaka onwards, minimum gas is 500.
+    }
+}
+
+template <Traits traits>
+uint64_t expmod_gas_cost(byte_string_view const input_view)
+{
+    constexpr uint64_t min_gas{expmod_min_gas<traits>()};
```

Comment:
I suppose the cost is fixed per revision, so:
```suggestion
    static constexpr uint64_t min_gas{expmod_min_gas<traits>()};
```